feat(docs): add TTLSecondsAfterFinished and RunPolicy to KEP-2899 (#3824)
kubeflow/training-operator: proposals/2899-resource-timeouts/README.md

### Goals

- Add `ActiveDeadlineSeconds` to `TrainJobSpec` for data scientists to control individual job timeouts
- Add `TTLSecondsAfterFinished` to `TrainJobSpec` for data scientists to control post-finish cleanup of individual jobs
- Add a `RunPolicy` struct to `TrainingRuntimeSpec` so platform admins can set default `ActiveDeadlineSeconds` and `TTLSecondsAfterFinished` for all TrainJobs using a runtime, with per-TrainJob overrides

> - Expose `ActiveDeadlineSeconds` and `TTLSecondsAfterFinished` in the Kubeflow Python SDK for data scientists

Reviewer: Let's add the goal that we should not break MultiKueue functionality when `TTLSecondsAfterFinished` is set.
Ref: https://cloud-native.slack.com/archives/C0742LDFZ4K/p1787858280074949

    // +kubebuilder:validation:Minimum=1
    // +kubebuilder:validation:XValidation:rule="self == oldSelf",message="field is immutable"

>     ActiveDeadlineSeconds *int64 `json:"activeDeadlineSeconds,omitempty"`

Reviewer: The shipped field is a **non-pointer** (`pkg/apis/trainer/v1alpha1/trainjob_types.go:139`) — `0` is the "unset" sentinel that `reconcileDeadline` gates on. Now that this section is labelled as shipped in v2.2, the snippet should match.

```suggestion
    ActiveDeadlineSeconds int64 `json:"activeDeadlineSeconds,omitempty"`
```

Reviewer: @XploY04 @robert-bell Actually, shall we use pointer here like in Job API?
https://github.com/kubernetes/api/blob/master/batch/v1/types.go#L330

We might want to update our TrainJob API too.

Reviewer 2: Yes we should be consistent with the Job api.

@andreyvelich this a breaking change to the api, and even though it's probably fairly safe I think we should keep it as is and address it when we promote to beta.

We need to get into the habit of avoiding breaking changes and working out how to deal with them.

@andreyvelich what do you think about tracking this with a `// todo(beta) replace with pointer` on the alpha api? https://github.com/kubeflow/trainer/blob/c71b47576050714b60a8d840916d8ad99bc1f81a/pkg/apis/trainer/v1alpha1/trainjob_types.go#L139

Do we have a tracking issue for promoting to beta? I can't find any what but I'm sure I've seen one.

Reviewer: >We need to get into the habit of avoiding breaking changes and working out how to deal with them.

Sure, we just discussed before with @tenzen-y that during the TrainJob alpha evolution we are ok to introduce breaking changes to move faster towards beta promotion.
But if you think that introducing all breaking changes once when we promote API to beta is better, we can also do this.

>Do we have a tracking issue for promoting to beta?

Not really, could you help to create one @robert-bell please?

Reviewer 2: Yeah for sure. I'll create it next week.

>     // TTLSecondsAfterFinished specifies the duration in seconds the TrainJob is retained after it
>     // reaches a terminal state (Complete or Failed), after which the TrainJob and its child
>     // resources become eligible for automatic deletion. Value must be a non-negative integer;
>     // 0 means delete immediately after finishing. Overrides the runtime default in
>     // spec.runPolicy.ttlSecondsAfterFinished. Following Kubernetes Job semantics, this field is
>     // mutable after creation.
>     // +optional
>     // +kubebuilder:validation:Minimum=0
>     TTLSecondsAfterFinished *int64 `json:"ttlSecondsAfterFinished,omitempty"`

Reviewer: Three fixes: use `*int32` (both `batch/v1.JobSpec` and `jobset.JobSetSpec` use `*int32` for this field — only `ActiveDeadlineSeconds` is `int64`), start the godoc with the lowercase JSON name per every field in `trainjob_types.go`, and make the field immutable (see the Validation comment below). Also `spec.runPolicy` reads as a TrainJob path in `kubectl explain` output.

```suggestion
    // ttlSecondsAfterFinished specifies the duration in seconds the TrainJob is retained after it
    // reaches a terminal state (Complete or Failed), after which the TrainJob and its child
    // resources become eligible for automatic deletion. Value must be a non-negative integer;
    // 0 means delete immediately after finishing. Overrides the referenced runtime's
    // spec.runPolicy.ttlSecondsAfterFinished default. The field is immutable.
    // +optional
    // +kubebuilder:validation:Minimum=0
    // +kubebuilder:validation:XValidation:rule="self == oldSelf",message="field is immutable"
    TTLSecondsAfterFinished *int32 `json:"ttlSecondsAfterFinished,omitempty"`
```

Reviewer 2: Why does it need to be immutable? Same question for active deadline.

Reviewer: Yeah, we can make it mutable. @XploY04 let's ensure we update it to *int32.

}
```

#### TrainingRuntimeSpec Changes

Add a `RunPolicy` struct to `TrainingRuntimeSpec` (used by both `ClusterTrainingRuntime` and `TrainingRuntime`) in `pkg/apis/trainer/v1alpha1/trainingruntime_types.go`:

```go
type TrainingRuntimeSpec struct {
    // ... existing fields ...

>     // RunPolicy defines lifecycle defaults applied to TrainJobs that reference this runtime.
>     // Each field is a default that a TrainJob can override on its own spec.
>     // +optional
>     RunPolicy *RunPolicy `json:"runPolicy,omitempty"`
> }
> 
> // RunPolicy holds runtime-level lifecycle defaults for TrainJobs using this runtime.
> type RunPolicy struct {
>     // ActiveDeadlineSeconds is the default maximum active duration in seconds for TrainJobs that
>     // reference this runtime. A TrainJob's own spec.activeDeadlineSeconds takes precedence.
>     // +optional
>     // +kubebuilder:validation:Minimum=1
>     ActiveDeadlineSeconds *int64 `json:"activeDeadlineSeconds,omitempty"`
> 
>     // TTLSecondsAfterFinished is the default retention duration in seconds after a TrainJob using
>     // this runtime reaches a terminal state, after which it is eligible for automatic deletion.
>     // A TrainJob's own spec.ttlSecondsAfterFinished takes precedence.
>     // +optional
>     // +kubebuilder:validation:Minimum=0
>     TTLSecondsAfterFinished *int64 `json:"ttlSecondsAfterFinished,omitempty"`
> }

Reviewer: Same conventions for the new struct: godoc starts with the lowercase JSON name, TTL is `*int32` and immutable, and `ActiveDeadlineSeconds` is non-pointer `int64` to match the shipped `TrainJobSpec` field.

```suggestion
    // runPolicy defines lifecycle defaults applied to TrainJobs that reference this runtime.
    // Each field is a default that a TrainJob can override on its own spec.
    // +optional
    RunPolicy *RunPolicy `json:"runPolicy,omitempty"`
}

// RunPolicy holds runtime-level lifecycle defaults for TrainJobs using this runtime.
type RunPolicy struct {
    // activeDeadlineSeconds is the default maximum active duration in seconds for TrainJobs that
    // reference this runtime. A TrainJob's own spec.activeDeadlineSeconds takes precedence.
    // +optional
    // +kubebuilder:validation:Minimum=1
    ActiveDeadlineSeconds int64 `json:"activeDeadlineSeconds,omitempty"`

    // ttlSecondsAfterFinished is the default retention duration in seconds after a TrainJob using
    // this runtime reaches a terminal state, after which it is eligible for automatic deletion.
    // A TrainJob's own spec.ttlSecondsAfterFinished takes precedence. The field is immutable.
    // +optional
    // +kubebuilder:validation:Minimum=0
    // +kubebuilder:validation:XValidation:rule="self == oldSelf",message="field is immutable"
    TTLSecondsAfterFinished *int32 `json:"ttlSecondsAfterFinished,omitempty"`
}
```

```

> The name `runPolicy` is used on the runtime (rather than a flat field) so it reads as policy applied to the resulting runs, not to the runtime object's own lifecycle. On `TrainJobSpec` the fields stay flat, matching the already-shipped `activeDeadlineSeconds`.

Reviewer: `runPolicy` is an existing `kubeflow.org/v1` type name from Training Operator v1, but there it lived on the job itself with per-job semantics and a different field set (`cleanPodPolicy`, `schedulingPolicy`, `backoffLimit`, `*int32` TTL). Worth one sentence noting the reuse is deliberate and the semantics differ (defaults on a template), so migrating v1 users don't assume v1 behaviour.

Reviewer: No need to mention this, since APIs are not backward compatible.

### Value Resolution

Both `ActiveDeadlineSeconds` and `TTLSecondsAfterFinished` follow the same precedence: the value on `TrainJobSpec` wins if set; otherwise the controller falls back to the runtime default in `TrainingRuntimeSpec.RunPolicy`; if neither is set, the feature is disabled for that TrainJob.

```
effectiveValue = trainJob.spec.<field>  (if set)
              else runtime.spec.runPolicy.<field>  (if set)
              else unset (no enforcement / no auto-cleanup)
```

> Resolution happens per reconcile, so a runtime default cannot silently override a value the user explicitly set on the TrainJob.

Reviewer: With `nil` meaning "inherit", a TrainJob can raise or lower a runtime default but never disable it — and for `activeDeadlineSeconds` there is no wire value for "no deadline" at all (non-pointer `int64` + `Minimum=1`, so `0` serializes away). That directly conflicts with the debugging scenario in Story 4 — sharper now that TTL is immutable, since the value can't be extended later either.

```suggestion
Resolution happens per reconcile, so a runtime default cannot silently override a value the user explicitly set on the TrainJob.

Because `nil` means "inherit", a TrainJob can raise or lower a runtime default but cannot disable it. For `activeDeadlineSeconds` the shipped field is a non-pointer `int64` with `Minimum=1`, so `0` is indistinguishable from unset and there is no wire value meaning "no deadline". For `ttlSecondsAfterFinished` the same limitation means a user cannot pin a failed run indefinitely for debugging once an admin sets a runtime TTL, and (given the field is immutable) cannot extend it after the fact either. State either that runtime defaults are an inescapable floor, or define an explicit opt-out.
```

kind: ClusterTrainingRuntime
metadata:
  name: torch-distributed-gpu
spec:
  runPolicy:
    activeDeadlineSeconds: 28800       # Default max runtime: 8 hours
    ttlSecondsAfterFinished: 3600      # Default cleanup: 1 hour after finish
  template:
    # ... JobSet template ...
```

**TrainJob overriding the runtime TTL default (Data Scientist):**

```yaml
apiVersion: trainer.kubeflow.org/v1alpha1
kind: TrainJob
metadata:
  name: keep-longer-for-debugging
spec:
  ttlSecondsAfterFinished: 86400       # Keep this failed run for 24h, overrides the 1h default
  runtimeRef:
    name: torch-distributed-gpu
  trainer:
    numNodes: 2
```

### Implementation Overview

**Controller Changes** (`pkg/controller/trainjob_controller.go`):

1. **Value Resolution:**
    - Resolve `ActiveDeadlineSeconds` and `TTLSecondsAfterFinished` using the precedence above: `TrainJobSpec` value if set, else `TrainingRuntimeSpec.RunPolicy` default, else unset

2. **Deadline Enforcement:**
    - Check if job is running and effective deadline is set
    - Calculate `deadline = startTime + effectiveActiveDeadlineSeconds` (where `startTime` is reset on each resume from suspension)

>     - If exceeded, mark TrainJob as Failed (`Reason: DeadlineExceeded`); the runtime framework handles cleanup of the underlying JobSet

Reviewer: `reconcileDeadline` deletes the JobSet inline in the controller (`pkg/controller/trainjob_controller.go:171-176`), not via the runtime framework.

```suggestion
    - If exceeded, mark TrainJob as Failed (`Reason: DeadlineExceeded`) and delete the child JobSet directly, as `reconcileDeadline` does today
```

    - Otherwise, requeue at `deadline`

3. **TTL Enforcement:**
    - Only applies once the TrainJob is finished (`Complete` or `Failed` condition is true)
    - Calculate `expiry = finishTime + effectiveTTLSecondsAfterFinished`, where `finishTime` is the transition time of the terminal condition

>     - If `expiry` has passed, delete the TrainJob; owner references cascade the delete to the child JobSet, Jobs, Pods, and Services, matching Kubernetes Job `ttlSecondsAfterFinished` semantics

Reviewer: The controller has no `delete` verb on `trainjobs` today (`pkg/controller/trainjob_controller.go:75` → `manifests/base/rbac/role.yaml:107-113` grants only `get,list,patch,update,watch`), so self-deletion would fail with a 403. Worth calling out explicitly.

```suggestion
    - If `expiry` has passed, delete the TrainJob; owner references cascade the delete to the child JobSet, Jobs, Pods, and Services, matching Kubernetes Job `ttlSecondsAfterFinished` semantics. This requires adding `delete` to the controller's TrainJob RBAC marker in `pkg/controller/trainjob_controller.go` (today `get;list;watch;update;patch`) and re-running `make generate`
```

**Field-level CEL markers** on the API types:

- `Minimum=1` on `ActiveDeadlineSeconds` (`TrainJobSpec` and `RunPolicy`)
- `Minimum=0` on `TTLSecondsAfterFinished` (`TrainJobSpec` and `RunPolicy`)

> - `XValidation: self == oldSelf` on `ActiveDeadlineSeconds` (`TrainJobSpec`) - immutable after creation
> - `TTLSecondsAfterFinished` is mutable after creation, matching Kubernetes Job behavior (users may extend or shorten retention on a finished job)

Reviewer: `TTLSecondsAfterFinished` should be **immutable** on both `TrainJobSpec` and `RunPolicy`, not mutable-per-Kubernetes-Job. Since the effective value is re-resolved on every reconcile, a mutable TTL lets a user re-arm or defer cleanup on an already-finished TrainJob.

```suggestion
- `XValidation: self == oldSelf` on `ActiveDeadlineSeconds` (`TrainJobSpec`) - immutable after creation
- `XValidation: self == oldSelf` on `TTLSecondsAfterFinished` (`TrainJobSpec` and `RunPolicy`) - immutable after creation. This intentionally diverges from Kubernetes Job, where TTL is mutable: because the effective value is re-resolved on every reconcile, a mutable TTL would let a user re-arm or defer cleanup on an already-finished TrainJob. Consequence on the runtime side: an admin cannot adjust `runPolicy.ttlSecondsAfterFinished` on an existing `ClusterTrainingRuntime` without recreating it
```

Reviewer 2: The fields on RunPolicy should be mutable. The snapshot mechanism means the runtime value isn't re-resolved on each reconcile.

Reviewer: Good point, I agree.

**Cross-field CEL markers** on `TrainingRuntimeSpec` to prevent conflicting lifecycle fields in the JobSet/Job template:

> - `self.template.spec.replicatedJobs.all(rj, !has(rj.template.spec.activeDeadlineSeconds))` - Job-level deadline would terminate pods independently from TrainJob deadline tracking

Reviewer: `spec.template` and `template.spec` are both optional in the shipped CRD (`spec` only carries `minProperties: 1`, no `required`), so an unguarded `self.template.spec...` raises `no such key` and rejects the runtime. This PR makes that reachable, since `runPolicy` alone now satisfies `minProperties: 1`. The rule also misses the Job-level field one level down.

```suggestion
- `!has(self.template) || !has(self.template.spec) || !has(self.template.spec.replicatedJobs) || self.template.spec.replicatedJobs.all(rj, !has(rj.template.spec.activeDeadlineSeconds))` - Job-level deadline would terminate pods independently from TrainJob deadline tracking. The `has()` guards are required because `spec.template` and `spec.template.spec` are both optional in the real schema, so an unguarded `self.template.spec.…` errors out and rejects runtimes that set only `runPolicy` and/or `mlPolicy`
```

- Deadline exceeded → job failed with DeadlineExceeded reason
- Deadline not reached → requeue at deadline
- Clock skew → requeue with delay instead of negative duration
- Value resolution: TrainJob value set → runtime default ignored
- Value resolution: TrainJob value unset, runtime default set → runtime default used
- Value resolution: neither set → feature disabled
- TTL on unfinished job → no deletion
- TTL expired on finished job → TrainJob deleted (children cascade)

> - TTL not reached on finished job → requeue at expiry

Reviewer: Missing the `0` boundary and the validation path. Please also add TTL + `runPolicy` cases to the integration test list below (it currently covers deadline only).

```suggestion
- TTL not reached on finished job → requeue at expiry
- `ttlSecondsAfterFinished: 0` on a finished job → deleted immediately
- Runtime template setting JobSet-level TTL → validation outcome asserted
```

- `test/e2e/trainjob_deadline_test.go`:
    - Real training workload with deadline: Verify job fails at timeout with DeadlineExceeded reason
    - Verify no orphaned resources remain
- `test/e2e/trainjob_ttl_test.go`:
    - Real training workload with TTL: verify the TrainJob and its child JobSet, Pods, and Services are deleted after the TTL elapses
    - Verify a failed TrainJob with maxRestarts stays terminally Failed and is not recreated after TTL cleanup (regression for [#3779](https://github.com/kubeflow/trainer/issues/3779))

## Future Plan

The initial `RunPolicy` struct holds `ActiveDeadlineSeconds` and `TTLSecondsAfterFinished`. If future lifecycle knobs (for example `BackoffLimit` or a cleanup policy) are needed, they can be added under `RunPolicy` without breaking changes.

## Implementation History

- **2025-10-20**: Issue opened [#2899](https://github.com/kubeflow/trainer/issues/2899)
- **2026-01-04**: Initial KEP drafted
- **2026-01-22**: KEP updated with layered API design (TrainJob + TrainingRuntime)

> - **2026-02-28**: `ActiveDeadlineSeconds` on `TrainJobSpec` shipped in v2.2

Reviewer: [#3258](https://github.com/kubeflow/trainer/pull/3258) merged 2026-03-14 and `v2.2.0` was tagged 2026-03-19.

```suggestion
- **2026-03-19**: `ActiveDeadlineSeconds` on `TrainJobSpec` shipped in v2.2.0 ([#3258](https://github.com/kubeflow/trainer/pull/3258))
```
